kernel/src/common/deferred_call_mux.rs

```diff
@@ -0,0 +1,207 @@
+//! ) };
+//! some_capsule.set_deferred_call_handle(
+//!     deferred_call_mux.register(some_capsule).expect("no deferred call slot available")
+//! );
+//! ```
+
+use crate::common::cells::OptionalCell;
+use crate::common::deferred_call::DeferredCall;
+use core::cell::Cell;
+
+pub const DEFERRED_CALL_MUX_TASK: usize = 31; // highest bit in usize reserved for deferred call mux
+
+static mut DEFERRED_CALL_MUX: Option<&'static DeferredCallMux> = None;
+
+/// Sets a global `DeferredCallMux` instance
+///
+/// This is required before any deferred calls can be retrieved.
+/// It may be called only once. Returns `true` if the mux was successfully
+/// registered.
+pub unsafe fn set_global_mux(mux: &'static DeferredCallMux) -> bool {
+    // If the returned reference is identical to the mux argument,
+    // it is set in the option. Otherwise, a different mux is
+    // already registered and may not be replaced.
+    (*DEFERRED_CALL_MUX.get_or_insert(mux)) as *const _ == mux as *const _
+}
+
+/// Call the globally registered mux
+///
+/// Returns true if a Mux was registered and has been called.
+/// This function needs to be called by the underlying deferred
+/// call implementation in the `chip` crate.
+pub unsafe fn call_global_mux() -> bool {
+    DEFERRED_CALL_MUX.map(|mux| mux.call()).is_some()
+}
+
+/// Multiplexer over [deferred calls](crate::common::deferred_call)
+///
+/// This multiplexer has a fixed number of possible clients, which
+/// is determined by the `clients`-array passed in with the constructor.
+pub struct DeferredCallMux {
+    deferred_call: DeferredCall<usize>,
+    clients: &'static [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>)],
+    handle_counter: Cell<usize>,
+}
+
+impl DeferredCallMux {
+    /// Construct a new deferred call
+    ///
+    /// This needs to be registered with the `set_global_mux` function immediately
+    /// afterwards, and should not be changed anymore. Only the globally registered
+    /// Mux will receive calls from the underlying deferred call implementation.
+    ///
+    /// The `clients` array can be initialized with any value. The recommended
+    /// values are (where n is the number of possible clients):
+    ///
+    /// `[(Cell::new(false), OptionalCell::empty()); n]`
+    pub unsafe fn new(
+        clients: &'static [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>)],
+    ) -> DeferredCallMux {
+        DeferredCallMux {
+            deferred_call: DeferredCall::new(DEFERRED_CALL_MUX_TASK),
+            clients: clients,
+            handle_counter: Cell::new(0),
+        }
+    }
+
+    /// Schedule a deferred call to be called
+    ///
+    /// The handle addresses the client that will be called.
+    ///
+    /// If no client for the handle is found (it was unregistered), this
+    /// returns `None`. If a call is already scheduled, it returns
+    /// `Some(false)`.
+    pub fn set(&self, handle: DeferredCallHandle) -> Option<bool> {
```

Comment:
This seems fine to me but wouldn't it be better to return `Result<(), Error>` here and use `Error::AlreadySet` and `Error::ClientNotRegistered`?

Returning `Some(false)`  made it a bit hard for me grasp at least

Reply:
Yes, I also think a `Result` better suits this function. However, I would then need to introduce yet another error enum for the `Mux` (e.g. something like `DeferredCallMuxSetCallError`, which is a bit crazy right?).

Also, do you think that `AlreadySet` should be considered an error? The caller is scheduled in this case, so I'm not quite sure what to think about this. I also don't think it's a good idea to hide this information as setting the handle twice could indicate a logic error.

Reply:
I think it is better to keep as it is for now, it is clean enough!

I think this belong in the conversation to create a  `chained result type` for all results in the kernel to encourage `Result` over `Option`